File: backend/app/schemas/practice.py
"""
Practice module - Pydantic schemas
"""
from pydantic import BaseModel, Field
from typing import Optional, List, Any
from uuid import UUID
from datetime import datetime


# Scenario schemas
class ScenarioBase(BaseModel):
    """Base scenario schema"""
    name: str = Field(..., max_length=100)
    description: Optional[str] = None
    type: str = Field(..., max_length=50)
    category: Optional[str] = Field(None, max_length=50)
    sub_category: Optional[str] = Field(None, max_length=50)
    default_role_config: dict = Field(default_factory=dict)


class ScenarioCreate(ScenarioBase):
    """Create scenario schema"""
    pass


class ScenarioUpdate(BaseModel):
    """Update scenario schema"""
    name: Optional[str] = Field(None, max_length=100)
    description: Optional[str] = None
    type: Optional[str] = Field(None, max_length=50)
    category: Optional[str] = Field(None, max_length=50)
    sub_category: Optional[str] = Field(None, max_length=50)
    default_role_config: Optional[dict] = None


class ScenarioResponse(ScenarioBase):
    """Scenario response schema"""
    id: UUID
    is_builtin: bool
    status: str
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class ScenarioListResponse(BaseModel):
    """Scenario list response schema"""
    data: List[ScenarioResponse]
    total: int


# Role config schemas
class RoleConfig(BaseModel):
    """Role configuration schema"""
    position_level: str = Field(..., description="岗位级别")
    personality: str = Field(..., description="性格特征")
    decision_style: str = Field(..., description="决策风格")


# Session schemas
class SessionCreate(BaseModel):
    """Create practice session schema"""
    scenario_id: UUID
    role_config: RoleConfig


class SessionResponse(BaseModel):
    """Session response schema"""
    id: UUID
    scenario_id: UUID
    user_id: UUID
    role_config: dict
    status: str
    started_at: Optional[datetime]
    ended_at: Optional[datetime]
    created_at: datetime

    class Config:
        from_attributes = True


class SessionCreateResponse(BaseModel):
    """Session create response with first message"""
    session_id: UUID
    websocket_url: str
    first_message: str


# Practice message schemas
class MessageResponse(BaseModel):
    """Practice message response"""
    id: UUID
    session_id: UUID
    role: str
    content: str
    audio_url: Optional[str]
    created_at: datetime

    class Config:
        from_attributes = True


# AI Client Role schemas
class AIClientRoleBase(BaseModel):
    """Base AI client role schema"""
    position_level: str = Field(..., max_length=50)
    personality: str = Field(..., max_length=50)
    decision_style: str = Field(..., max_length=50)
    context_window: Optional[str] = None


class AIClientRoleCreate(AIClientRoleBase):
    """Create AI client role schema"""
    pass


class AIClientRoleResponse(AIClientRoleBase):
    """AI client role response"""
    id: UUID
    created_at: datetime

    class Config:
        from_attributes = True


# Summary schemas
class SummaryCreate(BaseModel):
    """Create summary schema"""
    trigger: str = Field(..., description="auto or manual")
    rounds: int = Field(default=3, description="Number of rounds to summarize")


class SummaryResponse(BaseModel):
    """Summary response schema"""
    id: UUID
    session_id: UUID
    trigger_type: str
    rounds: str
    good_points: List[str]
    improvements: List[str]
    suggestions: List[str]
    positive_ratio: float
    created_at: datetime

    class Config:
        from_attributes = True


# Report schemas
class ReportResponse(BaseModel):
    """Report response schema"""
    id: UUID
    session_id: UUID
    communication_score: float
    persuasion_score: float
    closing_score: float
    spin_score: float
    summary: Optional[str]
    key_points: List[str]
    improvements: List[str]
    pdf_url: Optional[str]
    created_at: datetime

    class Config:
        from_attributes = True


# Session List schemas
class SessionListItem(BaseModel):
    """Session list item schema"""
    id: UUID
    scenario_id: UUID
    scenario_name: Optional[str] = None
    scenario_type: Optional[str] = None
    user_id: Optional[str] = None
    role_config: dict
    status: str
    current_phase: Optional[str] = None
    score: Optional[float] = None
    message_count: int = 0
    duration_minutes: int = 0
    created_at: datetime
    ended_at: Optional[datetime] = None

    class Config:
        from_attributes = True


class SessionListResponse(BaseModel):
    """Session list response schema"""
    data: List[SessionListItem]
    total: int
    page: int
    page_size: int


class DashboardStats(BaseModel):
    """Dashboard statistics schema"""
    total_sessions: int = 0
    total_time_minutes: int = 0
    avg_score: Optional[float] = None
    improvement_rate: float = 0.0
    this_week_sessions: int = 0
    this_week_time: int = 0
    this_week_score: Optional[float] = None
    streak_days: int = 0


class HistoryStats(BaseModel):
    """History statistics schema"""
    sessions: int = 0
    completed: int = 0
    duration_minutes: int = 0
    messages: int = 0
    avg_score: Optional[float] = None


class HistoryStatsResponse(BaseModel):
    """History stats response schema"""
    last_30_days: HistoryStats
    last_90_days: HistoryStats


# WebSocket message schemas
class WSUserMessage(BaseModel):
    """WebSocket user message"""
    type: str = "user_message"
    content: str
    timestamp: datetime = Field(default_factory=datetime.utcnow)


class WSAIMessage(BaseModel):
    """WebSocket AI message"""
    type: str = "ai_message"
    content: str
    audio_url: Optional[str] = None
    stream: bool = False
    timestamp: datetime = Field(default_factory=datetime.utcnow)

File: frontend/src/api/practice.ts
/**
 * Practice API Client
 */
import axios from 'axios'

const API_BASE = '/api/v1'

export interface Scenario {
  id: string
  name: string
  description: string
  type: string
  category: string
  sub_category: string
  default_role_config: object
  is_builtin: boolean
  status: string
  created_at: string
  updated_at: string
}

export interface RoleConfig {
  position_level: string
  personality: string
  decision_style: string
  voice: string  // TTS voice_id for AI customer
}

export interface SessionCreateResponse {
  session_id: string
  websocket_url: string
  first_message: string
}

// Session list types
export interface SessionListItem {
  id: string
  scenario_id: string
  scenario_name: string | null
  scenario_type: string | null
  user_id: string | null
  role_config: RoleConfig
  status: 'preparing' | 'active' | 'completed'
  current_phase: string | null
  score: number | null
  message_count: number
  duration_minutes: number
  created_at: string
  ended_at: string | null
}

export interface SessionListResponse {
  data: SessionListItem[]
  total: number
  page: number
  page_size: number
}

// Dashboard stats types
export interface DashboardStats {
  total_sessions: number
  total_time_minutes: number
  avg_score: number | null
  improvement_rate: number
  this_week_sessions: number
  this_week_time: number
  this_week_score: number | null
  streak_days: number
}

// History stats types
export interface HistoryStats {
  sessions: number
  completed: number
  duration_minutes: number
  messages: number
  avg_score: number | null
}

export interface HistoryStatsResponse {
  last_30_days: HistoryStats
  last_90_days: HistoryStats
}

// Session summary types
export interface SessionSummary {
  session_id: string
  overall_score: number
  situation_score: number
  problem_score: number
  implication_score: number
  need_payoff_score: number
  key_strengths: string[]
  areas_for_improvement: string[]
  next_practice_focus: string
}

// Phase summary types
export interface PhaseSummary {
  phase: string
  phase_label: string
  overall_score: number
  situation_score: number
  problem_score: number
  implication_score: number
  need_payoff_score: number
  good_points: string[]
  improvements: string[]
  suggestions: string[]
}

const api = {
  /**
   * 获取场景列表
   */
  async getScenarios(params?: {
    category?: string
    sub_category?: string
    type?: string
    is_builtin?: boolean
  }): Promise<{ data: Scenario[]; total: number }> {
    const response = await axios.get(`${API_BASE}/practice/scenarios`, { params })
    return response.data
  },

  /**
   * 获取单个场景
   */
  async getScenario(scenarioId: string): Promise<Scenario> {
    const response = await axios.get(`${API_BASE}/practice/scenarios/${scenarioId}`)
    return response.data
  },

  /**
   * 创建对练会话
   */
  async createSession(
    scenarioId: string,
    roleConfig: RoleConfig,
    customerContext?: any,
    investigationResult?: any,
    userContext?: any
  ): Promise<SessionCreateResponse> {
    const response = await axios.post(`${API_BASE}/practice/sessions`, {
      scenario_id: scenarioId,
      role_config: roleConfig,
      customer_context: customerContext,
      investigation_result: investigationResult,
      user_context: userContext,
    })
    return response.data
  },

  /**
   * 结束对练会话
   */
  async endSession(sessionId: string): Promise<void> {
    await axios.post(`${API_BASE}/practice/sessions/${sessionId}/end`)
  },

  /**
   * 获取会话总结报告
   */
  async getSessionSummary(sessionId: string): Promise<SessionSummary> {
    const response = await axios.get(`${API_BASE}/practice/sessions/${sessionId}/summary`)
    return response.data
  },

  /**
   * 下载会话报告 PDF（触发浏览器下载）
   */
  async downloadReportPdf(sessionId: string): Promise<void> {
    const response = await axios.get(
      `${API_BASE}/practice/sessions/${sessionId}/report/pdf`,
      { responseType: 'blob' }
    )
    const blob = new Blob([response.data], { type: 'application/pdf' })
    const url = window.URL.createObjectURL(blob)
    const a = document.createElement('a')
    a.href = url
    a.download = `salesagent-report-${sessionId.slice(0, 8)}.pdf`
    document.body.appendChild(a)
    a.click()
    document.body.removeChild(a)
    window.URL.revokeObjectURL(url)
  },

  /**
   * 获取会话列表（分页）
   */
  async getSessions(params?: {
    page?: number
    page_size?: number
    status?: string
    scenario_type?: string
  }): Promise<SessionListResponse> {
    const response = await axios.get(`${API_BASE}/practice/sessions-list`, { params })
    return response.data
  },

  /**
   * 获取仪表盘统计
   */
  async getDashboardStats(): Promise<DashboardStats> {
    const response = await axios.get(`${API_BASE}/practice/dashboard/stats`)
    return response.data
  },

  /**
   * 获取历史统计
   */
  async getHistoryStats(): Promise<HistoryStatsResponse> {
    const response = await axios.get(`${API_BASE}/practice/history/stats`)
    return response.data
  },

  /**
   * 获取单个会话详情
   */
  async getSession(sessionId: string): Promise<any> {
    const response = await axios.get(`${API_BASE}/practice/sessions/${sessionId}`)
    return response.data
  },

  /**
   * 获取阶段总结
   */
  async getPhaseSummary(sessionId: string, phase: string): Promise<PhaseSummary> {
    const response = await axios.post(`${API_BASE}/practice/sessions/${sessionId}/phases/${phase}/summary`)
    return response.data
  }
}

export default api